lib/Service/PhishingDetectionService.php

```diff
@@ -0,0 +1,190 @@
+		$this->trustedSenderService = $trustedSenderService;
+	}
+
+	private function checkDatePass(string $date): array {
+		$now = new DateTime();
+		$dt = new DateTime($date);
+		if($dt > $now) {
+			$this->warn = true;
+			return ["check" => false , "message" => "Sent date is in the future"];
+		}
+		return ["check" => $dt < $now ];
+	}
+
+	private function replyToCheckPass(string $fromEmail, ?string $replyToEmail): array {
+		if(!(isset($replyToEmail))) {
+			return ["check" => true];
+		}
+		if($replyToEmail !== $fromEmail) {
+			$this->warn = true;
+		}
+		return ["check" => false , "message" => "Reply-To email: ${$replyToEmail} is different from the sender email: ${$fromEmail}"];
+	}
+
+	private function customEmailCheck(string $fromEmail, ?string $customEmail): array {
+		if(!(isset($customEmail))) {
+			return ["check" => true];
+		}
+		if($customEmail !== $fromEmail) {
+			$this->warn = true;
+		}
+		return ["check" => false , "message" => "Sender is using a custom email: ${$customEmail} instead of the sender email: ${$fromEmail}"] ;
+	}
+
+
+	private function contactsCheckPass(string $fn, string $email):array {
+		$emailInContacts = false;
+		$emails = "";
+		$contacts = $this->contactIntegration->getContactsWithName($fn, true);
+		foreach ($contacts as $contact) {
+			foreach ($contact['email'] as $contactEmail) {
+				$emailInContacts = true;
+				$emails .= $contactEmail.",";
+				if ($contactEmail === $email) {
+					return ["check" => true];
+				}
+			}
+		}
+		if ($emailInContacts) {
+			$this->warn = true;
+			return ["check" => false, "message" => "Sender email: ${$email} is not in the contacts list, but the sender name: ${$fn} is in the contacts list with the following emails: ${$emails}"];
+		}
+		return ["check" => true];
+	}
+
+	private function checkTrusted(string $uid, string $email): array {
+		$domain = explode('@', $email)[1];
+		$trusted = $this->trustedSenderService->isTrusted($uid, $email) || $this->trustedSenderService->isTrusted($uid, $domain);
+
+		if(!$trusted) {
+			$this->warn = true;
+			return ["check" => false, "message" => "Sender email: ${$email} is not trusted"];
+		}
+		return ["check" => true];
+	}
+
+	private function isLink(string $text): bool {
+		$pattern = '/^(https?:\/\/|www\.|[a-zA-Z0-9-]+\.[a-zA-Z]{2,})/i';
+
+		return preg_match($pattern, $text) === 1;
+	}
+
+	private function getInnerText(\DOMElement $node) : string {
+		$innerText = '';
+		foreach ($node->childNodes as $child) {
+			if ($child->nodeType === XML_TEXT_NODE) {
+				$innerText .= $child->nodeValue;
+			} elseif ($child->nodeType === XML_ELEMENT_NODE) {
+				$innerText .= $this->getInnerText($child);
+			}
+		}
+		return $innerText;
+	}
+
+	private function parseAnchorTags(string $htmlMessage): array {
+
+		$results = [];
+		$zippedArray = [];
+
+		$dom = new \DOMDocument();
+		libxml_use_internal_errors(true);
+		$dom->loadHTML($htmlMessage);
+		libxml_use_internal_errors();
+		$anchors = $dom->getElementsByTagName('a');
```

Comment:
`area`-elements can contain links and could be checked, too. See https://developer.mozilla.org/en-US/docs/Web/API/Document/links

Reply:
So we would check href against alt in <area> ?

Reply:
Yes, or match the URL against a block-list.

Reply:
block-list is out of scope. 
I'll add a check against alt